File: packages/provide-uterm-server/src/provide/uterm/server/bridge/frames.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Typed-dict aliases and builder helpers for hijack-bridge wire frames.

The Pydantic ``schemas.py`` module in ``provide.uterm.bridge`` is the
single source of truth for frame shapes; builders here construct those
models and dump them so wire bytes stay byte-for-byte stable.

The local ``TypedDict`` aliases survive as the *return type* of each
builder so existing callers and downstream signatures don't have to
churn — they still see ``ErrorFrame``, ``TermFrame``, etc.
"""

from __future__ import annotations

import time
from typing import Any, Literal, TypedDict, cast

from provide.uterm.bridge.schemas import (
    AnalysisFrame as _AnalysisModel,
)
from provide.uterm.bridge.schemas import (
    ErrorFrame as _ErrorModel,
)
from provide.uterm.bridge.schemas import (
    HeartbeatAckFrame as _HeartbeatAckModel,
)
from provide.uterm.bridge.schemas import (
    HijackStateFrame as _HijackStateModel,
)
from provide.uterm.bridge.schemas import (
    PongFrame as _PongModel,
)
from provide.uterm.bridge.schemas import (
    TermFrame as _TermModel,
)
from provide.uterm.bridge.schemas import (
    WorkerConnectedFrame as _WorkerConnectedModel,
)
from provide.uterm.bridge.schemas import (
    WorkerDisconnectedFrame as _WorkerDisconnectedModel,
)
from provide.uterm.frames import make_snapshot_frame as _core_make_snapshot_frame


class ErrorFrame(TypedDict):
    type: Literal["error"]
    message: str


class PongFrame(TypedDict):
    type: Literal["pong"]
    ts: float


class HeartbeatAckFrame(TypedDict):
    type: Literal["heartbeat_ack"]
    lease_expires_at: float
    ts: float


class WorkerConnectedFrame(TypedDict):
    type: Literal["worker_connected"]
    worker_id: str
    ts: float


class WorkerDisconnectedFrame(TypedDict):
    type: Literal["worker_disconnected"]
    worker_id: str
    ts: float


class BrowserInputFrame(TypedDict):
    type: Literal["input"]
    data: str


class TermFrame(TypedDict):
    type: Literal["term"]
    data: str
    ts: float


class SnapshotFrame(TypedDict):
    type: Literal["snapshot"]
    screen: str
    cursor: dict[str, int]
    cols: int
    rows: int
    screen_hash: str
    cursor_at_end: bool
    has_trailing_space: bool
    prompt_detected: dict[str, Any] | None
    raw_tail: str | None
    ts: float


class AnalysisFrame(TypedDict):
    type: Literal["analysis"]
    formatted: str
    raw: Any
    ts: float


class HijackStateFrame(TypedDict):
    type: Literal["hijack_state"]
    hijacked: bool
    owner: str | None
    lease_expires_at: float | None
    input_mode: str


class WorkerStatusFrame(TypedDict, total=False):
    type: Literal["status"]
    ts: float


class HelloFrame(TypedDict, total=False):
    type: Literal["hello"]
    worker_id: str
    can_hijack: bool
    hijacked: bool
    hijacked_by_me: bool
    worker_online: bool
    input_mode: str
    role: str
    hijack_control: str
    hijack_step_supported: bool
    capabilities: dict[str, object]
    resume_supported: bool
    resume_token: str | None
    resumed: bool
    protocol_version: int
    # Range-negotiation handshake. Server hello sets
    # ``protocol={"selected": N, "server_min": MIN, "server_max": MAX}``.
    # Worker hello sets ``protocol={"min": ..., "max": ..., "preferred": ...}``.
    protocol: dict[str, int]


def make_error_frame(message: str) -> ErrorFrame:
    return cast("ErrorFrame", _ErrorModel(type="error", message=message).model_dump(exclude_none=True))


def make_pong_frame(*, ts: float | None = None) -> PongFrame:
    return cast(
        "PongFrame",
        _PongModel(type="pong", ts=time.time() if ts is None else ts).model_dump(exclude_none=True),
    )


def make_heartbeat_ack_frame(lease_expires_at: float, *, ts: float | None = None) -> HeartbeatAckFrame:
    return cast(
        "HeartbeatAckFrame",
        _HeartbeatAckModel(
            type="heartbeat_ack",
            lease_expires_at=lease_expires_at,
            ts=time.time() if ts is None else ts,
        ).model_dump(exclude_none=True),
    )


def make_worker_connected_frame(worker_id: str, *, ts: float | None = None) -> WorkerConnectedFrame:
    return cast(
        "WorkerConnectedFrame",
        _WorkerConnectedModel(
            type="worker_connected",
            worker_id=worker_id,
            ts=time.time() if ts is None else ts,
        ).model_dump(exclude_none=True),
    )


def make_worker_disconnected_frame(worker_id: str, *, ts: float | None = None) -> WorkerDisconnectedFrame:
    return cast(
        "WorkerDisconnectedFrame",
        _WorkerDisconnectedModel(
            type="worker_disconnected",
            worker_id=worker_id,
            ts=time.time() if ts is None else ts,
        ).model_dump(exclude_none=True),
    )


def make_term_frame(data: str, *, ts: float | None = None) -> TermFrame:
    return cast(
        "TermFrame",
        _TermModel(type="term", data=data, ts=time.time() if ts is None else ts).model_dump(exclude_none=True),
    )


def make_snapshot_frame(
    *,
    screen: str,
    cursor: dict[str, int],
    cols: int,
    rows: int,
    screen_hash: str,
    cursor_at_end: bool,
    has_trailing_space: bool,
    prompt_detected: dict[str, Any] | None,
    ts: float,
    raw_tail: str | None = None,
) -> SnapshotFrame:
    # Snapshot construction lives once in ``provide.uterm.frames``; this thin
    # wrapper only re-types the result as the local ``SnapshotFrame`` TypedDict
    # so server-side callers and signatures don't churn. The core builder
    # already preserves ``prompt_detected: None`` on the wire (exclude_none=False).
    return cast(
        "SnapshotFrame",
        _core_make_snapshot_frame(
            screen=screen,
            cursor=cursor,
            cols=cols,
            rows=rows,
            screen_hash=screen_hash,
            cursor_at_end=cursor_at_end,
            has_trailing_space=has_trailing_space,
            prompt_detected=prompt_detected,
            ts=ts,
            raw_tail=raw_tail,
        ),
    )


def make_analysis_frame(*, formatted: str, raw: Any, ts: float | None = None) -> AnalysisFrame:
    # ``raw`` is allowed to be None and must survive the dump — the
    # frontend reads it directly. Use exclude_none=False so a None ``raw``
    # is still serialised as ``"raw": null`` to match the legacy output.
    return cast(
        "AnalysisFrame",
        _AnalysisModel(
            type="analysis",
            formatted=formatted,
            raw=raw,
            ts=time.time() if ts is None else ts,
        ).model_dump(exclude_none=False),
    )


def make_hijack_state_frame(
    *,
    hijacked: bool,
    owner: str | None,
    lease_expires_at: float | None,
    input_mode: str,
) -> HijackStateFrame:
    # exclude_none=False — owner / lease_expires_at can legitimately be None
    # and the frontend reads them directly off the frame.
    return cast(
        "HijackStateFrame",
        _HijackStateModel(
            type="hijack_state",
            hijacked=hijacked,
            owner=owner,
            lease_expires_at=lease_expires_at,
            input_mode=input_mode,
        ).model_dump(exclude_none=False),
    )


def make_hello_frame(**payload: Any) -> HelloFrame:
    # Hello-frame payloads accept arbitrary capability flags that aren't
    # part of the Pydantic schema (``HelloFrame`` uses extra="ignore"),
    # so route through the loose TypedDict path and just stamp ``type``.
    return cast("HelloFrame", {"type": "hello", **payload})


def coerce_worker_status_frame(payload: dict[str, Any]) -> WorkerStatusFrame:
    frame = dict(payload)
    frame.setdefault("type", "status")
    frame.setdefault("ts", time.time())
    return cast("WorkerStatusFrame", frame)

File: packages/provide-uterm-go/frames/frames.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

// Package frames is a Go port of the WebSocket frame schema layer of
// provide-uterm. The single source of truth for wire shapes is the Pydantic
// module provide.uterm.bridge.schemas (packages/provide-uterm/src/provide/
// uterm/bridge/schemas.py); every struct here mirrors one model of the
// AnyFrame discriminated union (discriminator key: "type").
//
// Serialization semantics mirror Pydantic's model_dump(exclude_none=True):
// required fields are plain Go types and always serialize (including false
// booleans and zero numbers); optional (default-None) fields are pointers,
// or nil-able maps/slices/interfaces, tagged omitzero so absent values are
// omitted from the wire JSON.
//
// Extra-field policies from the Python models are honored by DecodeFrame:
//
//   - extra="forbid" (the default): unknown fields are rejected.
//   - extra="ignore" (HelloFrame): unknown fields are dropped.
//   - extra="allow" (IdentityFrame, StatusFrame, PresenceUpdateFrame,
//     PresenceSyncFrame): unknown fields round-trip via the Extra map.
//
// Deviation from Python: the server builders make_snapshot_frame,
// make_analysis_frame and make_hijack_state_frame dump with
// exclude_none=False, emitting explicit nulls for unset optionals
// (e.g. "prompt_detected": null, "raw": null, "owner": null). The Go
// marshalers uniformly use exclude_none=True semantics and omit those keys;
// consumers treat absent and null identically.
package frames

// Wire "type" discriminator literals, one per frame struct.
const (
	TypeTerm               = "term"
	TypeInput              = "input"
	TypeSnapshotReq        = "snapshot_req"
	TypeSnapshot           = "snapshot"
	TypeControl            = "control"
	TypeHijackState        = "hijack_state"
	TypeHijackRequest      = "hijack_request"
	TypeHijackRelease      = "hijack_release"
	TypeHijackStep         = "hijack_step"
	TypeWorkerConnected    = "worker_connected"
	TypeWorkerDisconnected = "worker_disconnected"
	TypeWorkerHello        = "worker_hello"
	TypeHeartbeat          = "heartbeat"
	TypeHeartbeatAck       = "heartbeat_ack"
	TypePing               = "ping"
	TypePong               = "pong"
	TypeHello              = "hello"
	TypeResume             = "resume"
	TypeIdentity           = "identity"
	TypeSessionToken       = "session_token"
	TypeResumeOk           = "resume_ok"
	TypeResumeFailed       = "resume_failed"
	TypeLinkPatterns       = "link_patterns"
	TypeAnalysis           = "analysis"
	TypeError              = "error"
	TypeStatus             = "status"
	TypeInputModeChanged   = "input_mode_changed"
	TypeApprovalPending    = "approval_pending"
	TypeApprovalResolved   = "approval_resolved"
	TypePresenceUpdate     = "presence_update"
	TypePresenceSync       = "presence_sync"
	TypePresenceLeave      = "presence_leave"
	TypeControlTransfer    = "control_transfer"
)

// IdentityFrame defaults (mirroring the Pydantic field defaults).
const (
	IdentityDefaultVersion     = 1
	IdentityDefaultFingerprint = ""
	IdentityDefaultTransport   = "ssh"
)

// ---------------------------------------------------------------------------
// Terminal data + snapshot
// ---------------------------------------------------------------------------

// TermFrame carries raw terminal output bytes from the worker to subscribers.
type TermFrame struct {
	Type string   `json:"type"`
	Data string   `json:"data"`
	TS   *float64 `json:"ts,omitzero"`
}

// InputFrame carries browser/operator input destined for the worker.
type InputFrame struct {
	Type string   `json:"type"`
	Data string   `json:"data"`
	TS   *float64 `json:"ts,omitzero"`
}

// SnapshotReqFrame is a browser-originated request for a fresh screen snapshot.
type SnapshotReqFrame struct {
	Type string   `json:"type"`
	TS   *float64 `json:"ts,omitzero"`
}

// SnapshotFrame is a worker-originated full-screen snapshot.
type SnapshotFrame struct {
	Type             string         `json:"type"`
	Screen           string         `json:"screen"`
	Cursor           map[string]int `json:"cursor,omitzero"`
	Cols             *int           `json:"cols,omitzero"`
	Rows             *int           `json:"rows,omitzero"`
	ScreenHash       *string        `json:"screen_hash,omitzero"`
	CursorAtEnd      *bool          `json:"cursor_at_end,omitzero"`
	HasTrailingSpace *bool          `json:"has_trailing_space,omitzero"`
	PromptDetected   map[string]any `json:"prompt_detected,omitzero"`
	// RawTail is a bounded rolling tail of raw decoded output (ANSI/control
	// intact) so consumers can recover content that scrolled off the 25-row
	// viewport within a single turn. Optional/backward-compatible.
	RawTail *string  `json:"raw_tail,omitzero"`
	TS      *float64 `json:"ts,omitzero"`
}

// ---------------------------------------------------------------------------
// Hijack lease lifecycle
// ---------------------------------------------------------------------------

// ControlFrame is a server-originated worker-control frame (pause/resume/step).
type ControlFrame struct {
	Type   string   `json:"type"`
	Action string   `json:"action"`
	Owner  *string  `json:"owner,omitzero"`
	LeaseS *float64 `json:"lease_s,omitzero"`
	TS     *float64 `json:"ts,omitzero"`
}

// HijackStateFrame is a broadcast lease-state update.
type HijackStateFrame struct {
	Type           string   `json:"type"`
	Hijacked       bool     `json:"hijacked"`
	Owner          *string  `json:"owner,omitzero"`
	LeaseExpiresAt *float64 `json:"lease_expires_at,omitzero"`
	InputMode      *string  `json:"input_mode,omitzero"`
}

// HijackRequestFrame is a browser-originated request to acquire the hijack lease.
type HijackRequestFrame struct {
	Type  string   `json:"type"`
	Token *string  `json:"token,omitzero"`
	TS    *float64 `json:"ts,omitzero"`
}

// HijackReleaseFrame is a browser-originated request to release the hijack lease.
type HijackReleaseFrame struct {
	Type string   `json:"type"`
	TS   *float64 `json:"ts,omitzero"`
}

// HijackStepFrame is a browser-originated single-step request.
type HijackStepFrame struct {
	Type string   `json:"type"`
	TS   *float64 `json:"ts,omitzero"`
}

// ---------------------------------------------------------------------------
// Worker presence
// ---------------------------------------------------------------------------

// WorkerConnectedFrame announces a worker connecting to the hub.
type WorkerConnectedFrame struct {
	Type     string   `json:"type"`
	WorkerID string   `json:"worker_id"`
	TS       *float64 `json:"ts,omitzero"`
}

// WorkerDisconnectedFrame announces a worker disconnecting from the hub.
type WorkerDisconnectedFrame struct {
	Type     string   `json:"type"`
	WorkerID string   `json:"worker_id"`
	TS       *float64 `json:"ts,omitzero"`
}

// WorkerHelloFrame is a worker-originated hello-frame carrying input_mode +
// capabilities.
type WorkerHelloFrame struct {
	Type string   `json:"type"`
	Mode *string  `json:"mode,omitzero"`
	TS   *float64 `json:"ts,omitzero"`
}

// ---------------------------------------------------------------------------
// Heartbeat / keepalive
// ---------------------------------------------------------------------------

// HeartbeatFrame is a browser-originated keepalive.
type HeartbeatFrame struct {
	Type string   `json:"type"`
	TS   *float64 `json:"ts,omitzero"`
}

// HeartbeatAckFrame is the server reply to a browser heartbeat — refreshes the
// lease.
type HeartbeatAckFrame struct {
	Type           string   `json:"type"`
	LeaseExpiresAt float64  `json:"lease_expires_at"`
	TS             *float64 `json:"ts,omitzero"`
}

// PingFrame is a liveness probe.
type PingFrame struct {
	Type string   `json:"type"`
	TS   *float64 `json:"ts,omitzero"`
}

// PongFrame is the reply to a PingFrame.
type PongFrame struct {
	Type string   `json:"type"`
	TS   *float64 `json:"ts,omitzero"`
}

// ---------------------------------------------------------------------------
// Hello / resume handshake
// ---------------------------------------------------------------------------

// HelloFrame is a server-originated hello-frame to the browser describing
// capabilities. The Python model is extra="ignore": unknown fields are
// silently dropped on decode (no Extra map).
type HelloFrame struct {
	Type                string         `json:"type"`
	WorkerID            *string        `json:"worker_id,omitzero"`
	CanHijack           *bool          `json:"can_hijack,omitzero"`
	Hijacked            *bool          `json:"hijacked,omitzero"`
	HijackedByMe        *bool          `json:"hijacked_by_me,omitzero"`
	WorkerOnline        *bool          `json:"worker_online,omitzero"`
	InputMode           *string        `json:"input_mode,omitzero"`
	Role                *string        `json:"role,omitzero"`
	HijackControl       *string        `json:"hijack_control,omitzero"`
	HijackStepSupported *bool          `json:"hijack_step_supported,omitzero"`
	Capabilities        map[string]any `json:"capabilities,omitzero"`
	ResumeSupported     *bool          `json:"resume_supported,omitzero"`
	McpSupported        *bool          `json:"mcp_supported,omitzero"`
	VncSupported        *bool          `json:"vnc_supported,omitzero"`
	ResumeToken         *string        `json:"resume_token,omitzero"`
	Resumed             *bool          `json:"resumed,omitzero"`
	ProtocolVersion     *int           `json:"protocol_version,omitzero"`
	// Protocol carries the range-negotiation handshake. Server hello sets
	// {"selected": N, "server_min": MIN, "server_max": MAX}; worker hello
	// sets {"min": ..., "max": ..., "preferred": ...}.
	Protocol map[string]int `json:"protocol,omitzero"`
	TS       *float64       `json:"ts,omitzero"`
}

// ResumeFrame asks the server to resume a previous session.
type ResumeFrame struct {
	Type     string `json:"type"`
	Token    string `json:"token"`
	PlayerID *int   `json:"player_id,omitzero"`
}

// IdentityFrame is the inline control-channel identity frame. The Python
// model is extra="allow": unknown fields round-trip via Extra. Version,
// Fingerprint and Transport have Pydantic defaults (1, "", "ssh") which
// NewIdentityFrame applies and DecodeFrame restores when the keys are absent.
type IdentityFrame struct {
	Type        string         `json:"type"`
	Version     int            `json:"version"`
	Subject     string         `json:"subject"`
	Fingerprint string         `json:"fingerprint"`
	Transport   string         `json:"transport"`
	Claims      map[string]any `json:"claims,omitzero"`
	Signature   *string        `json:"signature,omitzero"`
	// Extra holds unknown wire fields (extra="allow").
	Extra map[string]any `json:"-"`
}

// SessionTokenFrame hands the browser a resumable session token.
type SessionTokenFrame struct {
	Type     string `json:"type"`
	Token    string `json:"token"`
	PlayerID *int   `json:"player_id,omitzero"`
}

// ResumeOkFrame acknowledges a successful resume.
type ResumeOkFrame struct {
	Type string `json:"type"`
}

// ResumeFailedFrame reports a failed resume attempt.
type ResumeFailedFrame struct {
	Type   string  `json:"type"`
	Reason *string `json:"reason,omitzero"`
}

// LinkPatternEntry is one clickable-link pattern inside a LinkPatternsFrame.
// It is not itself a frame (no "type" discriminator). The Python field
// "class_" serializes under its alias "class". Action must be one of
// "cmd", "url", "key" or "focus".
type LinkPatternEntry struct {
	Pattern string  `json:"pattern"`
	Action  string  `json:"action"`
	ID      *string `json:"id,omitzero"`
	Flags   *string `json:"flags,omitzero"`
	// Group is int | string in the Python schema.
	Group   any     `json:"group,omitzero"`
	Payload any     `json:"payload,omitzero"`
	Hover   *string `json:"hover,omitzero"`
	// LineContains scopes the pattern: the client only runs the regex on
	// lines whose (SGR-stripped) text contains this substring.
	// Empty/absent = every line.
	LineContains *string `json:"line_contains,omitzero"`
	Class        *string `json:"class,omitzero"`
}

// LinkPatternsFrame pushes the clickable-link pattern set to the browser.
type LinkPatternsFrame struct {
	Type     string             `json:"type"`
	Patterns []LinkPatternEntry `json:"patterns"`
}

// ---------------------------------------------------------------------------
// Misc server → browser
// ---------------------------------------------------------------------------

// AnalysisFrame carries a formatted analysis blob plus its raw payload.
type AnalysisFrame struct {
	Type      string   `json:"type"`
	Formatted string   `json:"formatted"`
	Raw       any      `json:"raw,omitzero"`
	TS        *float64 `json:"ts,omitzero"`
}

// ErrorFrame reports a bridge error. Protocol-mismatch close frames carry the
// reason/min/max extras.
type ErrorFrame struct {
	Type      string  `json:"type"`
	Message   string  `json:"message"`
	Reason    *string `json:"reason,omitzero"`
	ClientMin *int    `json:"client_min,omitzero"`
	ClientMax *int    `json:"client_max,omitzero"`
	ServerMin *int    `json:"server_min,omitzero"`
	ServerMax *int    `json:"server_max,omitzero"`
}

// StatusFrame is a worker-originated status passthrough. The Python model is
// extra="allow": the type discriminator and ts are the only modelled fields;
// arbitrary status payloads round-trip via Extra.
type StatusFrame struct {
	Type string   `json:"type"`
	TS   *float64 `json:"ts,omitzero"`
	// Extra holds unknown wire fields (extra="allow").
	Extra map[string]any `json:"-"`
}

// InputModeChangedFrame announces a change of the worker's input mode.
type InputModeChangedFrame struct {
	Type      string   `json:"type"`
	InputMode string   `json:"input_mode"`
	TS        *float64 `json:"ts,omitzero"`
}

// ---------------------------------------------------------------------------
// Approval gating
// ---------------------------------------------------------------------------

// ApprovalPendingFrame announces a command awaiting operator approval.
type ApprovalPendingFrame struct {
	Type      string  `json:"type"`
	Command   string  `json:"command"`
	RequestID string  `json:"request_id"`
	ExpiresAt float64 `json:"expires_at"`
}

// ApprovalResolvedFrame announces the outcome of an approval request.
type ApprovalResolvedFrame struct {
	Type      string `json:"type"`
	Outcome   string `json:"outcome"`
	RequestID string `json:"request_id"`
}

// ---------------------------------------------------------------------------
// Presence (DeckMux wire format)
// ---------------------------------------------------------------------------

// PresenceUpdateFrame is a DeckMux per-user presence update. The Python model
// is extra="allow" because optional fields (scroll, selection, pin, typing,
// queued_keys) are attached only when relevant — they round-trip via Extra.
type PresenceUpdateFrame struct {
	Type   string  `json:"type"`
	UserID *string `json:"user_id,omitzero"`
	// Extra holds unknown wire fields (extra="allow").
	Extra map[string]any `json:"-"`
}

// PresenceSyncFrame is the full presence roster sent on browser connect.
// The Python model is extra="allow".
type PresenceSyncFrame struct {
	Type    string           `json:"type"`
	Users   []map[string]any `json:"users,omitzero"`
	Config  map[string]any   `json:"config,omitzero"`
	OwnerID *string          `json:"owner_id,omitzero"`
	// Extra holds unknown wire fields (extra="allow").
	Extra map[string]any `json:"-"`
}

// PresenceLeaveFrame announces a user leaving the presence roster.
type PresenceLeaveFrame struct {
	Type   string   `json:"type"`
	UserID string   `json:"user_id"`
	TS     *float64 `json:"ts,omitzero"`
}

// ControlTransferFrame is a DeckMux ownership-transfer notice.
type ControlTransferFrame struct {
	Type       string  `json:"type"`
	FromUserID *string `json:"from_user_id,omitzero"`
	ToUserID   *string `json:"to_user_id,omitzero"`
	Reason     *string `json:"reason,omitzero"`
	QueuedKeys *string `json:"queued_keys,omitzero"`
}
